nav2_controller/src/nav2_controller.cpp

```diff
@@ -385,11 +388,34 @@ void ControllerServer::computeAndPublishVelocity()
 
   nav_2d_msgs::msg::Twist2D twist = getThresholdedTwist(odom_sub_->getTwist());
 
-  auto cmd_vel_2d =
-    controllers_[current_controller_]->computeVelocityCommands(
-    pose,
-    nav_2d_utils::twist2Dto3D(twist),
-    goal_checker_.get());
+  geometry_msgs::msg::TwistStamped cmd_vel_2d;
+
+  if (controller_patience_ > 0 || controller_patience_ == -1.0) {
```

Comment:
You can remove the `else`, and move the `if` condition inside of the `catch()`. That way you only have 1 compute velocity call and if it catches, that's when you care about the patience to `throw()` if exceeding. I think all of the patience logic can be put into `catch() {...}`, you don't want to even reach the publish stage if you're going to be throwing a failure

Reply:
I moved all the patience logic inside the `catch() {...}` and piped the controller exception in case we don't use patience. Plus added the `publishVelocity` 0 inside (I assume we want to have the last cmd published to be 0 in case of exception). However, we now loose feedback when the controller plugin throws exceptions that we catch with the patience.
Plus properly getting the parameter and corrected the hardcoded 5 seconds patience (in a word, I properly tested and debuged it)

Reply:
Woah - if you want `publishVelocity`, when what's wrong with the current fail -> retry mechanism in the BT via the recovery nodes? I thought the goal here was to remove any stops. I believe this is what Nav1 did, but I can't recall off the top of my head

Reply:
We'll also have to deal with that feedback issue - but lets deal with the big issues first

Reply:
> Though I'm reading it, and I think it will still do feedback and publish velocities. You `catch` but if you don't rethrow due to patience failure, there's no `return`, so you will continue in to that last code block it looks like to me! (which also brings up the issue of not zeroing out the cmd_vel command outputted in case it's non-zero, you also wouldn't need to `publishZeroVelocity()` in the catch either if we go down to a `publishVelocity`. If we rethrow it due to being over time, then the main `catch` in the server will terminate the action and send the zero velocity command itself)

You are right, no need for the added `publishVelocity` and there is no feedback issue (try / catch manages to confuse me even when I write the code). Fixed in last commit (+ proper dealing with `controller_patience` set to -1.0 for infinite patience)